src/server/auth/server/api_server.go

```diff
@@ -750,57 +814,79 @@ func (a *apiServer) ModifyAdmins(ctx context.Context, req *auth.ModifyAdminsRequ
 	// Canonicalize GitHub usernames in request (must canonicalize before we can
 	// validate, so we know who is actually being added/removed & can confirm
 	// that not all admins are being removed)
-	eg := &errgroup.Group{}
-	canonicalizedToAdd := make([]string, len(req.Add))
-	for i, user := range req.Add {
-		i, user := i, user
-		eg.Go(func() error {
-			user, err = a.canonicalizeSubject(ctx, user)
-			if err != nil {
-				return err
-			}
-			canonicalizedToAdd[i] = user
-			return nil
-		})
+	canonicalizedUser, err := a.canonicalizeSubject(ctx, req.Principal)
+	if err != nil {
+		return nil, err
 	}
-	canonicalizedToRemove := make([]string, len(req.Remove))
-	for i, user := range req.Remove {
-		i, user := i, user
-		eg.Go(func() error {
-			user, err = a.canonicalizeSubject(ctx, user)
-			if err != nil {
-				return err
+
+	// Check if the current modify request contains a SUPER role for this principal
+	var grantSuper bool
+	var grantFS bool
+	if req.Roles != nil {
+		for _, role := range req.Roles.Roles {
+			if role == auth.AdminRole_SUPER {
+				grantSuper = true
 			}
-			canonicalizedToRemove[i] = user
-			return nil
-		})
-	}
-	if err := eg.Wait(); err != nil {
-		return nil, err
+			if role == auth.AdminRole_FS {
+				grantFS = true
+			}
+		}
 	}
-	err = a.validateModifyAdminsRequest(canonicalizedToAdd, canonicalizedToRemove)
+
+	err = a.validateModifyAdminsRequest(canonicalizedUser, grantSuper)
 	if err != nil {
 		return nil, err
 	}
 
+	// Get the user's existing grants
+	var hasSuper bool
+	var hasFS bool
+	func() {
```

Comment:
Rewrote this particular instance to do everything based on etcd state in a single STM. I think this line of thinking also extends to all of the admin interface - we currently use the cache to check whether an operation will remove all SUPER admins, but we read from the cache so it's entirely possible to interleave operations like:

- 2 admins, A and B
- RPC to remove A checks the cache, sees B
- RPC to remove B checks the cache, sees A
- A is removed from etcd
- B is removed from etcd
- no admins exist

Do we have a sense of how expensive it would be to just eliminate the cache altogether? I'm thinking along the lines of "ListAdmins should just go to the source, but then we have no good way to interrogate the state of the cache".

Reply:
I looked into it briefly, listing keys doesn't seem to be supported in STM so I'm not sure how we could remove the race in checking for at least one admin. One possibility would be to keep a separate integer key we increment and decrement as part of the transaction. Another would be to reorganize the data into a single, large key with all admins. Either of those seems beyond the scope of this PR.

Reply:
Yeah this is a really good point on both counts (that we shouldn't use the cache to validate modifyAdmins and that fixing this will be hard).

I haven't finished review the new commits yet, but I think what you already did (fixing this instance, since it can be done in an STM easily, and ignoring admin validation for now) makes complete sense to me.

(For 1.12, it seems like it's definitely time to implement auth extract/restore, which would let us change how admins are represented. I like the idea of having a single large key for all admins—it would amount to a cluster-level ACL, which is consistent with the GCP-ization of the auth model, where the cluster is a resource and SUPER and FS are the roles with respect to that resource.

Finally, if I'm being honest, part of me thinks the cache might still have value for `Authorize()` and `WhoAmI` since they get called so much—every RPC, and a single top-level user RPC might trigger a lot of internal RPCs—but a) I haven't tested it, and b) there are definitely safer ways to implement that. E.g. making sure the cache is only used in Authorize and WhoAmI, and nowhere else, or caching it but keying the cache with an ID tied to the top-level request, or embeddingsome of the caller's resource roles, including admin status, in the RPC metadata and propagating that for internal requests).